main.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"time"
+
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+type serviceResource struct {
+	name     string
+	stype    corev1.ServiceType
+	port     int32
+	selector map[string]string
+	ns       string
+}
+
+type k8sServiceAPI struct {
+	cs kubernetes.Interface
+	s  *serviceResource
+}
+
+const ns string = "mynamespace"
+
+var api *k8sServiceAPI
+
+func main() {
+	api = &k8sServiceAPI{
+		cs: getKubeHandle(),
+	}
+
+	watchlist := cache.NewListWatchFromClient(api.cs.AppsV1().RESTClient(), "deployments",
+		ns, fields.Everything())
+	_, controller := cache.NewInformer(watchlist, &v1.Deployment{}, time.Second*0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    onAdd,
+			DeleteFunc: onDelete})
+
+	stop := make(chan struct{})
+	go controller.Run(stop)
+	for {
+		time.Sleep(time.Second)
+	}
+}
+
+func onAdd(obj interface{}) {
+	dep := obj.(*v1.Deployment)
+	name := dep.GetName()
+
+	if name == "" {
+		return
+	}
+	service, err := extractServiceInfoFromDeployment(dep)
+	if err != nil {
+		return
+	}
+	api.s = service
+	exists := isServiceExists(api)
```

Comment:
let's try to make this mode readable, by making a controller struct and have client fields in that controller struct.

Reply:
new controller added